// n8n-workflow/dist/src/WorkflowHooks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class WorkflowHooks {
    constructor(hookFunctions, mode, executionId, workflowData, optionalParameters) {
        optionalParameters = optionalParameters || {};
        this.hookFunctions = hookFunctions;
        this.mode = mode;
        this.executionId = executionId;
        this.workflowData = workflowData;
        this.sessionId = optionalParameters.sessionId;
        this.retryOf = optionalParameters.retryOf;
    }
    async executeHookFunctions(hookName, parameters) {
        if (this.hookFunctions[hookName] !== undefined && Array.isArray(this.hookFunctions[hookName])) {
            for (const hookFunction of this.hookFunctions[hookName]) {
                await hookFunction.apply(this, parameters)
                    .catch((error) => {
                    console.error(`There was a problem executing hook: "${hookName}"`);
                    console.error('Parameters:');
                    console.error(parameters);
                    console.error('Error:');
                    console.error(error);
                });
            }
        }
    }
}
exports.WorkflowHooks = WorkflowHooks;
